In the implementation of the "reopen-log" ejabberdctl command,
do not attempt to rename log files, appending the suffix "-old"
to their base names, before opening them again.
(This behaviour breaks the "create" option of logrotate
which instructs logrotate to re-create the rotated log files.)

Also this patch replaces module for handling SASL error reporting
by a custom one which opens its log file in append mode rather
than in rewrite mode as the default (ERTS) module does
which makes a call to `ejabberdctl reopen-log` truncate
the erlang.log file.

Index: pkg-ejabberd/src/ejabberd_logger_h.erl
===================================================================
--- pkg-ejabberd.orig/src/ejabberd_logger_h.erl	2010-05-11 22:25:45.000000000 +0400
+++ pkg-ejabberd/src/ejabberd_logger_h.erl	2010-05-11 22:25:51.000000000 +0400
@@ -82,7 +82,6 @@
     remove_handler;
 handle_info({emulator, _GL, reopen}, State) ->
     file:close(State#state.fd),
-    rotate_log(State#state.file),
     case file:open(State#state.file, [append, raw]) of
 	{ok, Fd} ->
 	    {ok, State#state{fd = Fd}};
Index: pkg-ejabberd/src/ejabberd_admin.erl
===================================================================
--- pkg-ejabberd.orig/src/ejabberd_admin.erl	2010-05-11 22:25:45.000000000 +0400
+++ pkg-ejabberd/src/ejabberd_admin.erl	2010-05-11 22:26:45.000000000 +0400
@@ -210,9 +210,8 @@
     ejabberd_logger_h:reopen_log(),
     case application:get_env(sasl,sasl_error_logger) of
 	{ok, {file, SASLfile}} ->
-	    error_logger:delete_report_handler(sasl_report_file_h),
-	    ejabberd_logger_h:rotate_log(SASLfile),
-	    error_logger:add_report_handler(sasl_report_file_h,
+	    error_logger:delete_report_handler(sasl_report_append_file_h),
+	    error_logger:add_report_handler(sasl_report_append_file_h,
 	        {SASLfile, get_sasl_error_logger_type()});
 	_ -> false
 	end,
Index: pkg-ejabberd/src/sasl_report_append_file_h.erl
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ pkg-ejabberd/src/sasl_report_append_file_h.erl	2010-05-11 22:28:43.000000000 +0400
@@ -0,0 +1,47 @@
+%%
+%% SASL report logging to a file in append mode.
+%% Inspired by Erlang's sasl_report_file_h module.
+%%
+-module(sasl_report_append_file_h).
+
+%%%
+%%% A handler that can be connected to the error_logger
+%%% event handler.
+%%% Writes all sasl_* events formatted to file
+%%%
+
+-export([init/1,
+	 handle_event/2, handle_call/2, handle_info/2,
+	 terminate/2]).
+
+init({File, Type}) ->
+    process_flag(trap_exit, true),
+    case file:open(File, [append]) of
+	{ok,Fd} ->
+	    {ok, {Fd, File, Type}};
+	What ->
+	    What
+    end.
+
+handle_event({_Type, GL, _Msg}, State) when node(GL) /= node() ->
+    {ok, State};
+handle_event(Event, {Fd, File, Type}) ->
+    sasl_report:write_report(Fd, Type, tag_event(Event)),
+    {ok, {Fd, File, Type}};
+handle_event(_, State) ->
+    {ok, State}.
+
+handle_info({'EXIT', Fd, _Reason}, {Fd, _File, _Type}) ->
+    remove_handler;
+handle_info(_, State) ->
+    {ok, State}.
+
+handle_call(_Query, _State) -> {error, bad_query}.
+
+terminate(_, {Fd, _File, _Type}) ->
+    file:close(Fd),
+    [].
+
+tag_event(Event) ->
+    {calendar:local_time(), Event}.
+
